Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIWebRequest.mm

```diff
@@ -53,6 +53,254 @@
+
+    [manager loadAndRun];
+
+    EXPECT_NS_EQUAL(manager.get().yieldMessage, @"Load Tab");
+
+    [manager.get().defaultTab.mainWebView loadRequest:urlRequest];
+
+    [manager run];
+}
+
+TEST(WKWebExtensionAPIWebRequest, DeniedFilterTest)
+{
+    TestWebKitAPI::HTTPServer server({
+        { "/"_s, { { { "Content-Type"_s, "text/html"_s } }, ""_s } },
+    }, TestWebKitAPI::HTTPServer::Protocol::Http);
+
+    auto *backgroundScript = Util::constructScript(@[
+        // Setup
+        @"function passListener() { browser.test.notifyPass() }",
+        @"function failListener() { browser.test.notifyFail('This listener should not have been called') }",
+
+        // The passListener firing (but not the failListener) will consider the test passed.
+        @"browser.webRequest.onCompleted.addListener(failListener, { 'urls': [ '*://*.example.com/*' ] })",
+        @"browser.webRequest.onCompleted.addListener(passListener)",
+
+        // Yield after creating the listener so we can load a tab.
+        @"browser.test.yield('Load Tab')"
+    ]);
+
+    auto extension = adoptNS([[_WKWebExtension alloc] _initWithManifestDictionary:webRequestManifest resources:@{ @"background.js": backgroundScript }]);
+    auto manager = adoptNS([[TestWebExtensionManager alloc] initForExtension:extension.get()]);
+
+    // Grant the webRequest permission.
+    [manager.get().context setPermissionStatus:_WKWebExtensionContextPermissionStatusGrantedExplicitly forPermission:_WKWebExtensionPermissionWebRequest];
+
+    auto *urlRequest = server.requestWithLocalhost();
+    auto *matchPattern = [_WKWebExtensionMatchPattern matchPatternWithString:[NSString stringWithFormat:@"*://*.%@/*", urlRequest.URL.host]];
+    [manager.get().context setPermissionStatus:_WKWebExtensionContextPermissionStatusGrantedExplicitly forMatchPattern:matchPattern];
+
+    [manager loadAndRun];
+
+    EXPECT_NS_EQUAL(manager.get().yieldMessage, @"Load Tab");
+
+    [manager.get().defaultTab.mainWebView loadRequest:urlRequest];
+
+    [manager run];
+}
+
+TEST(WKWebExtensionAPIWebRequest, AllEventsFiredTest)
+{
+    TestWebKitAPI::HTTPServer server({
+        { "/"_s, { { { "Content-Type"_s, "text/html"_s } }, ""_s } },
+    }, TestWebKitAPI::HTTPServer::Protocol::Http);
+
+    auto *backgroundScript = Util::constructScript(@[
+        // Setup
+        @"let beforeRequestFired = false",
+        @"let beforeSendHeadersFired = false",
+        @"let sendHeadersFired = false",
+        @"let headersReceivedFired = false",
+        @"let responseStartedFired = false",
+        @"function beforeRequestHandler() { beforeRequestFired = true }",
+        @"function beforeSendHeadersHandler() { beforeSendHeadersFired = true }",
+        @"function sendHeadersHandler() { sendHeadersFired = true }",
+        @"function headersReceivedHandler() { headersReceivedFired = true }",
+        @"function responseStartedHandler() { responseStartedFired = true }",
```

Comment:
These would be clearer inline with `addListener` below.

```
browser.webRequest.onBeforeRequest.addListener(() => { beforeRequestFired = true })
```

Reply:
Way clearer.